File: Source/Remote Control Server/WhitelistWindow.vb
Public Class WhitelistWindow

    Public Sub validateInput()
        Dim countValid As Integer = 0
        Dim countInvalid As Integer = 0
        Dim whitelist As String = tb_whitelist.Text
        For Each ip As String In whitelist.Split(vbNewLine)
            If Network.isValidIp(ip.Trim) Then
                countValid += 1
            Else
                countInvalid += 1
            End If
        Next

        Dim status As String
        If countValid = 1 Then
            status = countValid & " valid IP address found"
        Else
            status = countValid & " valid IP addresses found"
        End If
        If countInvalid > 0 Then
            status = status & ", " & countInvalid & " invalid"
            btn_apply.Enabled = False
        Else
            btn_apply.Enabled = True
        End If
        label_count.Text = status
    End Sub

    Private Sub WhitelistWindow_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        For Each ip As String In Settings.whitelistedIps
            tb_whitelist.Text = tb_whitelist.Text & ip & vbNewLine
        Next

        'Remove last linebreak
        tb_whitelist.Text = tb_whitelist.Text.Substring(0, tb_whitelist.Text.LastIndexOf(vbNewLine))

        validateInput()
    End Sub

    Private Sub tb_whitelist_TextChanged(sender As Object, e As EventArgs) Handles tb_whitelist.TextChanged
        validateInput()
    End Sub

    Private Sub btn_apply_Click(sender As Object, e As EventArgs) Handles btn_apply.Click
        Settings.whitelistedIps = New List(Of String)
        Dim whitelist As String = tb_whitelist.Text
        For Each ip As String In whitelist.Split(vbNewLine)
            If Network.isValidIp(ip.Trim) Then
                If Not Settings.whitelistedIps.Contains(ip.Trim) Then
                    Settings.whitelistedIps.Add(ip.Trim)
                End If
            End If
        Next
        Me.Close()
    End Sub

End Class

File: Source/MacOS/RemoteControlServer/whitelistwindow.cpp
#include "network.h"
#include "settings.h"
#include "whitelistwindow.h"
#include "ui_whitelistwindow.h"

#include <QDebug>

WhitelistWindow::WhitelistWindow(QWidget *parent) :
    QDialog(parent),
    ui(new Ui::WhitelistWindow)
{
    ui->setupUi(this);

    connect(ui->ipTextEdit, SIGNAL(textChanged()), this, SLOT(validateInput()));
    connect(ui->applyButton, SIGNAL(pressed()), this, SLOT(customHide()));
}

WhitelistWindow::~WhitelistWindow()
{
    delete ui;
}

void WhitelistWindow::customHide()
{
    Settings::Instance()->whitelistedIps = new QStringList();
    QString whitelist = ui->ipTextEdit->toPlainText();
    foreach(QString ip, whitelist.split("\n"))
    {
        QString trimmedIp = ip.trimmed();
        if (Network::Instance()->isValidIp(trimmedIp))
        {
            if (!Settings::Instance()->whitelistedIps->contains(QString(ip.trimmed())))
            {
                Settings::Instance()->whitelistedIps->append(QString(ip.trimmed()));
            }
        }
    }
    hide();
}

void WhitelistWindow::customShow()
{
    QString text = "";
    for (int i = 0; i < Settings::Instance()->whitelistedIps->length(); ++i)
    {
        text += Settings::Instance()->whitelistedIps->at(i) + "\n";
    }
    text = text.left(text.length() - 1);
    ui->ipTextEdit->setText(text);

    validateInput();

    show();
}

void WhitelistWindow::validateInput()
{
    int countValid = 0;
    int countInvalid = 0;
    QString whitelist = ui->ipTextEdit->toPlainText();

    foreach (QString ip, whitelist.split("\n"))
    {
        QString trimmedIp = ip.trimmed();
        if (Network::Instance()->isValidIp(trimmedIp)) countValid += 1;
        else countInvalid += 1;
    }

    QString status;
    if (countValid == 1) status = QString::number(countValid) + " valid IP address found";
    else status = QString::number(countValid) + " valid IP addresses found";

    if (countInvalid > 0)
    {
        status = status + ", " + QString::number(countInvalid) + " invalid";
        ui->applyButton->setEnabled(false);
    }
    else
    {
        ui->applyButton->setEnabled(true);
    }
    ui->validIPsFound->setText(status);
}
